### PR TITLE
Refactor device pool

### PR DESCRIPTION
**Pull request by ShakoHo**
resend the pull request
@zapion , @ypwalter r?

### CODE REVIEW

File: utils/device_pool.py [hunk 1]
```diff
@@ -1,107 +1,211 @@
...
-
-    def _check_serial_in_devices(self):
-        if self.serial:
-            chk_list = filter(lambda x: x == self.serial, self.devices)
-            if not chk_list:
-                logger.warning("Android serial[" + self.serial + "] can't be found")
-                return False
-            else:
-                return True
+def exec_process(cmd_list):
```
Comment (zapion) [inline utils/device_pool.py]: quick question, what is the benefit for wrapping popen.communicate?

File: utils/device_pool.py [hunk 2]
```diff
@@ -1,107 +1,211 @@
...
+
+    def remove_adb_forward(self):
+        # Remove port forwarding
+        ret = os.system("ADNDROID_SERIAL=" + self.serial + " adb forward --remove tcp:" + str(self.adb_forwarded_port))
+        if ret != 0:
+            raise DMError("can't forward port to ANDROID_SERIAL[" + self.serial + "]")
+
+
+class DevicePool(object):
+
```
Comment (zapion) [inline utils/device_pool.py]: Probably re-using upper-object makes it more elegant?
Then you can avoid writing
import device_pool.DevicePool
and turn to use
import device_pool

File: utils/device_pool.py (hunk 1 above)
Comment (ShakoHo) [inline utils/device_pool.py]: the main purpose is to close the stdin, stdout and stderr obj, I found some ppl are discussion the OSError 24 is caused by not closing these objs

File: utils/device_pool.py (hunk 2 above)
Comment (ShakoHo) [inline utils/device_pool.py]: Sorry...but I don't get your point about re-using upper-object?...Could you talk about more detail?

File: utils/device_pool.py (hunk 1 above)
Comment (zapion) [inline utils/device_pool.py]: Please help to add comment to address this issue, thanks.

File: utils/device_pool.py [hunk 3]
```diff
@@ -1,107 +1,211 @@
...
             return
         try:
-            self.my_lock.release()
+            self.current_device_object.release_file_lock()
         except NotLocked:
-            logger.info("Lock is released")
-        self.my_lock = None
+            logger.info("File lock is released")
+
+        if self.current_device_object.is_in_forwarded_list:
```
Comment (zapion) [inline utils/device_pool.py]: should be a function call?

File: utils/device_pool.py [hunk 4]
```diff
@@ -1,107 +1,211 @@
...
+    def _find_available_port(self):
+        os.environ["ANDROID_SERIAL"] = self.serial
+        if 'ANDROID_SERIAL' in os.environ.keys():
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sock.bind(('', 0))
+            port = sock.getsockname()[1]
+            sock.close()
+            return port
+
+    def is_file_locked(self):
```
Comment (zapion) [inline utils/device_pool.py]: Should be a property?

File: utils/device_pool.py [hunk 5]
```diff
@@ -1,107 +1,211 @@
...
+        return self.lock_file.i_am_locking()
+
+    def is_in_forwarded_list(self):
```
Comment (zapion) [inline utils/device_pool.py]: Should be a property?

File: utils/device_pool.py [hunk 6]
```diff
@@ -1,107 +1,211 @@
...
+            ret = subprocess.call('/usr/bin/adb forward --list |grep ' + self.serial, shell=True, stdout=open(os.devnull, 'w'))
+            if ret == 0:
+                return True
+            else:
+                return False
+        else:
+            logger.error("adb forward --list not supported; recommend to upgrade 1.0.31 or newer version")
+            return False
+
+    def get_adb_forwarded_port(self):
```
Comment (zapion) [inline utils/device_pool.py]: Should be a property?

File: utils/device_pool.py (hunk 4 above)
Comment (ypwalter) [inline utils/device_pool.py]: @property ++

File: utils/device_pool.py (hunk 3 above)
Comment (ypwalter) [inline utils/device_pool.py]: If Shako changes this function to property, it should be fine. However, line 192 and 195 and other lines needed to be handled.

File: utils/device_pool.py [hunk 7]
```diff
@@ -1,107 +1,208 @@
...
+        return True
+
+    def remove_adb_forward(self):
+        # Remove port forwarding
+        ret = os.system("ADNDROID_SERIAL=" + self.serial + " adb forward --remove tcp:" + str(self.adb_forwarded_port))
+        if ret != 0:
+            raise DMError("can't forward port to ANDROID_SERIAL[" + self.serial + "]")
+
+
+def _gen_serial_list():
```
Comment (zapion) [inline utils/device_pool.py]: Probably a public function; gen_device_obj_map() as well

**Comment by zapion** [comment]
Most of modification look great!

**Comment by zapion** [comment]
r+, thanks!

**Pull request merged**